Introduce Lcc analysis

ricardarosemann opened the pull request:
## Purpose of this PR
- This PR adds a another reporting function and a plotting functionality to provide the following analysis:
  - Estimate the lifetime of heating systems
  - Based on the lifetime, compute the lifecycle costs (LCC).
  - Compute the levelized costs of heating (LCOH).
  - From the LCC, estimate the heating system shares assuming a logit model and compare to the actual Brick shares.
  - Plot the Brick renovation flow ratios w.r.t gas as dependent on the LCC difference to gas to visualise the price sensitivity.

## Implementation changes
- Additional function ```reportLCCShare``` to write a csv file containing all the data from the analysis.
- Several functions called from ```reportLCCShare``` to compute each step of the analysis
  - For the lifetime estimate, three different methods are available, of which only two are still used.
- Generalize the function ```plotBRICKCalib``` to ```plotBRICKAdditional``` which can generate both the calibration and the LCC analysis report.
- Add a function to normalize the price sensitivity; this is called in ```reportLCCShare```, but the result is not written to the csv or used otherwise (yet).
- Additionally: Adjust the ```checkMatching``` function to the renaming of ```reg``` to ```region```.

## Remarks
- I got a cyclocomp linter warning on ```showAnalysisPlot```, which means that this functions has too many different paths leading through it (given by if-clauses and switches). I ignored it for the time being and I actually like the function the way it is. But maybe we still find a cleverer way to handle this.
- I have introduced a lot of new files with the new functions. I'm a bit unsure whether we should try to reduce the number of files and/or functions. In general, I am very open to any suggestions on the structure of the code.
- I'm also now thinking I could move the smaller subfunctions inside the main function of the respective file. What do you think?
- I considered vectorizing the ```reportLCCShare``` function, similar to the calibration and calibration reporting. Since it is not necessary (yet?) and would involve quite some work, I decided against it for the moment. But happy to hear your thoughts.

ricardarosemann committed:
Implement lifetime computation as ex-post, ex-ante and mixed; create plots

ricardarosemann committed:
Implement lifetime computation as ex-post, ex-ante and mixed; create plots

ricardarosemann committed:
Compute lifetimes, LCC, logit and brick shares, introduce
plotBRICKAdditional

Implement lifetime and LCC, LCOH computations. Introduce computation of logit and brick shares.
Logit shares are weighted with brick renovation initial systems.
Generalize plotBRICKCalib to plotBRICKAdditional, enable this to render lifetime and LCC plots.
Add first implementation of price sensitivity normalization.

ricardarosemann committed:
Compute lifetimes, LCC, logit and brick shares, introduce
plotBRICKAdditional

Implement lifetime and LCC, LCOH computations. Introduce computation of logit and brick shares.
Logit shares are weighted with brick renovation initial systems.
Generalize plotBRICKCalib to plotBRICKAdditional, enable this to render lifetime and LCC plots.
Add first implementation of price sensitivity normalization.

ricardarosemann committed:
Restructure computation of LCC and brick shares: Do not aggregate across hs. Remove energy ladder filtering in the computation

ricardarosemann committed:
Restructure computation of LCC and brick shares: Do not aggregate across hs. Remove energy ladder filtering in the computation

ricardarosemann committed:
Adjust LCC computation to up-to-date brick

Rename reg to region, adjust standing lifetime, switch to simplified
ex-ante lifetime computation for consistency.

Add plots and flexibilize plotting function.

ricardarosemann committed:
Adjust LCC computation to up-to-date brick

Rename reg to region, adjust standing lifetime, switch to simplified
ex-ante lifetime computation for consistency.

Add plots and flexibilize plotting function.

ricardarosemann committed:
Prepare implementation for hs x hsr cost plots

Write renovation LCC and LCOH to the csv at full resolution, implement
three options to visualise by hs and hsr as functions.
So far no plots have beend added to the pdf output.

ricardarosemann committed:
Prepare implementation for hs x hsr cost plots

Write renovation LCC and LCOH to the csv at full resolution, implement
three options to visualise by hs and hsr as functions.
So far no plots have beend added to the pdf output.

ricardarosemann committed:
Shorter and extended LCC analysis

Introduce separate plotting routines for a more condensed and an
extended LCC analysis.

Add line plots with remaining share of heating systems, include status
quo bias in cost computation, adjust optics of the plot and
implementation structure.

ricardarosemann committed:
Shorter and extended LCC analysis

Introduce separate plotting routines for a more condensed and an
extended LCC analysis.

Add line plots with remaining share of heating systems, include status
quo bias in cost computation, adjust optics of the plot and
implementation structure.

ricardarosemann committed:
Scale LCC for logit share computation, tidy up

Introduce function to normalize LCC by lifetime and use the normalized
LCC in logit share compuation.
Tidy up all functions of the lifetime, LCC and LCOH analysis.
Rename ttot to ttotIn and ttot2 to ttotOut, remove unnecessary dt
column.

ricardarosemann committed:
Scale LCC for logit share computation, tidy up

Introduce function to normalize LCC by lifetime and use the normalized
LCC in logit share compuation.
Tidy up all functions of the lifetime, LCC and LCOH analysis.
Rename ttot to ttotIn and ttot2 to ttotOut, remove unnecessary dt
column.

ricardarosemann committed:
LCC: Remove dependency of Weibull params, restructure share plot

Tidy up LCC analysis functionalities:
All computations no longer require the Weibull parameters directly, but
use the shares computed in brick.
Plotting functionalities have been moved from the Rmd files to an actual
function of the package.

Restructure brick/logit share plot:
Plot heating system shares by previous heating system
for the full resolution data.
Move the hs x hsr matrix plots to the plotting function and slightly
adjust heat map plot.

ricardarosemann committed:
LCC: Remove dependency of Weibull params, restructure share plot

Tidy up LCC analysis functionalities:
All computations no longer require the Weibull parameters directly, but
use the shares computed in brick.
Plotting functionalities have been moved from the Rmd files to an actual
function of the package.

Restructure brick/logit share plot:
Plot heating system shares by previous heating system
for the full resolution data.
Move the hs x hsr matrix plots to the plotting function and slightly
adjust heat map plot.

ricardarosemann committed:
Adapt LCC analysis to seqRen, fix bugs, prettify plots

Fix small bugs
Join column in LCOH computation was wrong, is now fixed.
LCC plots now take 'valueMax' argument, anything surpassing it is cut
from the plot and marked by transparent colors (supported for single bar
plot and matrix charts).

ricardarosemann committed:
Adapt LCC analysis to seqRen, fix bugs, prettify plots

Fix small bugs
Join column in LCOH computation was wrong, is now fixed.
LCC plots now take 'valueMax' argument, anything surpassing it is cut
from the plot and marked by transparent colors (supported for single bar
plot and matrix charts).

ricardarosemann committed:
Linearized relation LCC with ratio, tidy up showAnalysisPlot

Make LCC analysis runnable after rebase
Implementation of the linearized relation between the LCC
and quantity ratios.
Tidy up showAnalysisPlot
Handle more than one region for the LCC analysis by introducing a
regionHandling parameter to plotBRICKAdditional.

ricardarosemann committed:
Linearized relation LCC with ratio, tidy up showAnalysisPlot

Make LCC analysis runnable after rebase
Implementation of the linearized relation between the LCC
and quantity ratios.
Tidy up showAnalysisPlot
Handle more than one region for the LCC analysis by introducing a
regionHandling parameter to plotBRICKAdditional.

ricardarosemann committed:
Fix linter warnings in LCC analysis

... and delete unused functions and code bits.

ricardarosemann committed:
Fix linter warnings in LCC analysis

... and delete unused functions and code bits.

ricardarosemann commented:
Closes https://github.com/pik-piam/reportbrick/issues/11

ricardarosemann committed:
Add flexibility for LCC plots

Allow to fix the y-axis for bar plots, update handling of valueCap in bar
plots.
Move the valueCap argument to the optional plot arguments.

ricardarosemann committed:
Add flexibility for LCC plots

Allow to fix the y-axis for bar plots, update handling of valueCap in bar
plots.
Move the valueCap argument to the optional plot arguments.

ricardarosemann committed:
Make lcc runnable with Brick version 0.12.2

Account for the additional 'vin' dimension in p_shareRen..Init
parameters.

ricardarosemann committed:
Make lcc runnable with Brick version 0.12.2

Account for the additional 'vin' dimension in p_shareRen..Init
parameters.

ricardarosemann committed:
Flexibilize reference for cost differences, write to file

In lcc analyis:
Allow to set any heating system as the reference in
computations for the linearized logistic model. Write the LCC
differences to a separate csv file.
In share plots, no longer omit hydrogen.

ricardarosemann committed:
Flexibilize reference for cost differences, write to file

In lcc analyis:
Allow to set any heating system as the reference in
computations for the linearized logistic model. Write the LCC
differences to a separate csv file.
In share plots, no longer omit hydrogen.

ricardarosemann committed:
Make showAnalyisPlot more stable

ricardarosemann committed:
Make showAnalyisPlot more stable

ricardarosemann committed:
Adapt to changed upstream Brick, improve documentation

Enable lcc-analysis to handle changed p_shareRenHSinit parameters,
preserve backwards compatibility.

Structure the main script a bit more and add comments.

ricardarosemann committed:
Adapt to changed upstream Brick, improve documentation

Enable lcc-analysis to handle changed p_shareRenHSinit parameters,
preserve backwards compatibility.

Structure the main script a bit more and add comments.

ricardarosemann committed:
Add plots of time evolution of LCOH

ricardarosemann committed:
Add plots of time evolution of LCOH

ricardarosemann commented:
I pushed an updated version of the LCC-analysis. An example of the current output is here: 
[BRICK_analysis_report_reg_DEU.pdf](https://github.com/user-attachments/files/27004961/BRICK_analysis_report_reg_DEU.pdf)

This includes some simplifications in the implementation and some more plots on the levelized costs of heat.

I think it would make sense to discuss the resulting plots first, before you dive more deeply into the implementation.
The pdf is quite long and I'm unsure how to shorten it. It might make sense to also compute weighted averages of the reported variables to reduce the number of plots. But for diagnostic purposes I still believe it makes sense to keep the possibility of generating the full output.

The code also includes some remnants of an earlier approach to the lifetime computation and other earlier trials. I haven't removed them yet, but most likely will do so before the merge.

ricardarosemann committed:
Clean LCC-analysis and adjust plots

Remove `dims` as argument to various function.
Flexibilize `.aggregateShare`.
Aggregate data across vintages.
Consistently filter full resolution data for specific vintage or ttotIn.
Remove unnecessary plots.
Switch/modify plots as discussed, in particular introduce plots with
aggregation across vintages.

ricardarosemann committed:
Remove unused code from LCC analysis

Remove ex-post lifetime computation.
Remove LCC/LCOH computed from ex-ante lifetimes.
Remove rescaled LCCs.
Remove Rmd-file of extended plots.

ricardarosemann commented:
I again pushed an updated version of the LCC-analysis, including a rebase on the current state of `reportbrick`. I added two commits, including two major points we discussed @robinhasse:
- Adjust plots as discussed and modify data where necessary.
- Remove all unnecessary code.

One major step still remains to be done in my view:
- Renaming of the files and more importantly, of all variables. My proposal would be to switch to the style of the `mif`-file, but to stick to the `.csv`-output file for the time being, with additional columns to depict the higher dimensionality. Then e.g. the renovation LCC resolved by hs, hsr, loc and typ could be named as "LCC|Renovation|hs|hsr|loc|typ". But this is just a suggestion, I'm happy to hear your thoughts (and maybe discuss offline). The information on which hs, hsr and so on would still be in the extra columns. Long term we might want to fully switch to the `.mif`-syntax and maybe also to the `mip`-plotting routines. But in my view this would be too involved for now, and the high dimensionality of the data makes this tedious.

Nevertheless, this is already usable. The standard procedure to obtain the LCC-plots consists of two steps:
- Call `reportLCCShare(<path-to-Brick-folder-with-output.gdx-file>)`. If necessary, `gdxName` can be set to conduct the analysis on a gdx file that is not called `output.gdx`.
- Call `plotBRICKAdditional(<path-to-Brick-folder-with-output-from-reportLCCShare>, plottingRoutine = "plotsLcc.Rmd")`. This generates a separate pdf-file for each region (and therefore takes quite a while). You can select a single region by setting `regionHandling = <regioncode>`.

robinhasse commented:
Thanks Ricarda, really great work overall. I'm just collecting thoughts here. None of these are showstoppers. Maybe just inspiration for future development.
- even with fixed buildings, the tool generate seemingly meaningful plots for construction. This could be misleading. 
- pie charts can get too wide for DIN A4 page
- in LCOH renovation plots, the BRICK share scale has different limits sometimes reaching beyond 1 (which I guess should not be reached?!). We could set fixed limits from 0 to 1.
- The x-axis label in the LCOH renovation plots could be changed to "final heating system" to help get the direction right.
- I find the costType label could be improved to show that intangible and tangible refer to investment cost.
- What are semi-transparent bas in the LCOH time evolution?
- Why don't we see end effects in the LCOH time evolution for HPs?

ricardarosemann commented:
>* even with fixed buildings, the tool generate seemingly meaningful plots for construction. This could be misleading.
>
Good point. I'll try to suppress that; this shouldn't be too involved.
> * pie charts can get too wide for DIN A4 page
>
Never happened for me, I think - this I will skip because I don't really know how to manipulate the pdf page setup. Also, I still plan to eventually switch to html, then this should not be necessary anymore.
>* in LCOH renovation plots, the BRICK share scale has different limits sometimes reaching beyond 1 (which I guess should not be reached?!). We could set fixed limits from 0 to 1.
>
This has to do with how I scale the secondary axis. Most likely it still shouldn't be too complicated to fix the secondary axis, but I'd also postpone this for the time being.
>* The x-axis label in the LCOH renovation plots could be changed to "final heating system" to help get the direction right.
>
Good point and hopefully an easy fix, thanks!
>* I find the costType label could be improved to show that intangible and tangible refer to investment cost.
>
I'll also try to do that, sounds reasonable (and easy).
>* What are semi-transparent bas in the LCOH time evolution?
>
That's when the actual value is cut off because it exceeds some maximum - I'm using this in several places, if one entry is much larger than others, so that the other bars can be evaluated better. Not sure on how to make this more transparent though.
>* Why don't we see end effects in the LCOH time evolution for HPs?
>
Most likely because heat pumps have the shortest life times and I already cut off three time periods at the end (i.e. 15 years).

ricardarosemann committed:
Adjust plots according to suggestions

... and remove further unused code.

ricardarosemann committed:
Rename functions and variables of LCC analysis

Adjust the naming logic of intermediate and output variables of LCC
analysis.
Rename key functions and the output files.

ricardarosemann commented:
As discussed, I included some further adjustments:
- Rename the internal variables in the code.
- Rename the output variables of the reporting function and thus the variables printed to the plots.
- Rename the reporting function itself as well as some further functions and subfunctions; I hope I consistently also renamed the files, if not, I'm happy to correct that.
- Rename the output csv and pdf files.

An example of the resulting pdf can be seen here: 
[BRICK_Heating_System_report_reg_DEU.pdf](https://github.com/user-attachments/files/28811315/BRICK_Heating_System_report_reg_DEU.pdf)

Some caveats remain:
- I did not yet implement a separate line the plot headings to specify over which dimension(s) I aggregated (as we had discussed); this information is currently contained in the variable name.
- Variables tend to be rather long, so that I had to introduce a line break in the heading in case there are two variables in the plot.
- Long variable names in some case lead to rather narrow plots, when the variable name has to be written into the legend.
- In case of the lifetime plots the meaning of solid versus dashed line changes (i.e. in lifetime plots "dashed" is the maximum lifetime, while in the the remaining share plot, "dashed" is the lifetime estimate from Brick) - this happens because the order of the variables is currently determined by the alphabet. Perspectively, this should be adjusted.
- `showAnalysisPlot` does not yet support line plots with only the color aesthetic and not the linetype aesthetic.

Feel free to have a look at this now and if you like, decide whether you already want to merge or wait for these caveats to be fixed.
